reference_retrieval_api.md

```diff
@@ -0,0 +1,424 @@
+
+| Error type             | HTTP status code | Description                                                                                          |
+|------------------------|------------------|------------------------------------------------------------------------------------------------------|
+| `Bad Request`            | 400              | Cannot process due to malformed request, the requested parameters do not adhere to the specification |
+| `Unauthorized`           | 401              | Authorization provided is invalid                                                                    |
+| `Not Found`              | 404              | The resource requested was not found                                                                 |
+| `Unsupported Media Type` | 415              | The requested sequence formatting is not supported by the server                                     |
+| `Range Not Satisfiable`  | 416              | The Range request cannot be satisfied                                                                |
+| `Not Implemented`        | 501              | The specified request is not supported by the server                                                 |
+
+## Security
+The reference sequence retrieval API can enable the retrieval of potentially sensitive genomic data by means of a client/server model. Effective security measures are essential to protect the integrity and confidentiality of these data.
+
+Sensitive information transmitted on public networks, such as access tokens and human genomic data, MUST be protected using Transport Level Security (TLS) version 1.2 or later, as specified in [RFC 5246](https://tools.ietf.org/html/rfc5246).
+
+If the data holder requires client authentication and/or authorization, then the client’s HTTPS API request MUST present an OAuth 2.0 bearer access token as specified in [RFC 6750](https://tools.ietf.org/html/rfc6750), in the Authorization request header field with the Bearer authentication scheme:
+
+```
+Authorization: Bearer [access_token]
+```
+
+The policies and processes used to perform user authentication and authorization, and the means through which access tokens are issued, are beyond the scope of this API specification. GA4GH recommends the use of the OAuth 2.0 framework ([RFC 6749](https://tools.ietf.org/html/rfc6749)) for authentication and authorization.
+
+## Checksum calculation
+The supported checksum algorithms are `MD5` and a SHA-512 based system called `TRUNC512` (see later for details). Servers MUST support sequence retrieval by one or more of these algorithms, and are encouraged to support all to maximize interoperability. To provide CRAM Reference Registry compatibility an implementation must support MD5.
+
+When calculating the checksum for a sequence, all non-base symbols (\n, spaces, etc) must be removed and then uppercase the rest. The allowed alphabet for checksum calculation is uppercase ASCII (0x41-0x5A).
```

Comment:
This is different to the SAM specification of how MD5 is computed; see SAM spec "Reference MD5 calculation" section.

Specifically it strips out any character of 32 (space) or below and 127 (delete) and above, and then uppercases.  Specifically this means punctuation remains as-is.

I don't know whether that is particularly sensible, but it's what the code actually did  so the spec was written to match it!  The practical situation is that it makes no difference for most data, but occasionally there may be something else.

I wonder if we also need to consider the methylation work being done elsewhere which uses 1,2,3,4 and 7 characters as well as lowercase letters (c vs C differ).  This breaks things for everyone, but the API could be set up with an additional parameter that requests case sensitivity?  (As long as there is room to add this, it doesn't need to go in now.)

Reply:
We did discuss this on the call with @daviesrob and from re-reading the minutes I think these are the pertinent points.

- The original htslib code was pointed out but that for the most part those other characters are treated as `N`. The sum result was restrict to just A-Z and have encodings treat this as N (unspecified in the spec).
- Main source of data for the RR API service are INSDC submitted sequences, which restrict to `(?i)^([ACGTUMRWSYKVHDBN]+)\\*?$"`. Far more restrictive than what we've said here in the spec
- The VMC spec, of which `TRUNC512` was based on restricts to just IUPAC nucleotide or amino acid seqs
- All source algorithms agreed on the uppercasing of sequence before being digested

That resulted in the decision to allow only upper-cased `A-Z` and this maintained API compatibility with CRAM RR not with htslib. 

Given the other examples I'm not sure how some of the more obscure sequence would make it into the ENA implementation of this but I am somewhat worried now. @lairdm thoughts?

As for the methylation that really begins to mess with the ideas behind what sequence is :). I'd nearly be tempted to suggest this is an alternative checksum algorithm. What do others think?

Reply:
That's a fair comment - if the existing reference server has a set of additional restrictions over and above the SAM specification, then the difference is largely irrelevant.

Yes we can find data where the M5 tag in SAM mismatches the MD5 calculation by this API, but such references couldn't be used externally due to other restrictions.  Seems fine then.

Reply:
Do you feel it's worth adding a condensed version of this into the specification? I know the VMC does this to explain key design decisions to inform future implementors/maintainers.

Reply:
I think a brief rationale could help.

Reply:
Brief rationale added